src/devtools/client/inspector/markup/new-markup.js

```diff
@@ -0,0 +1,261 @@
+  }
+
+  destroy() {
+    this.inspector.sidebar.off("markupview-selected", this.update);
+    this.selection.off("new-node-front", this.update);
+
+    this.inspector = null;
+    this.nodes = null;
+    this.selection = null;
+    this.store = null;
+    this.tree = null;
+  }
+
+  /**
+   * Given the NodeFront, add a representation of the node's properties to the
+   * markup tree object.
+   *
+   * @param  {NodeFront} node
+   *         The NodeFront of the node to add to the markup tree.
+   * @param  {Boolean} isExpanded
+   *         Whether or not the node is expanded.
+   */
+  async addNodeToTree(node, { isExpanded = false } = {}) {
+    const parentNode = node.parentNode();
+    const id = node.objectId();
+
+    this.tree[id] = {
+      // A list of the node's attributes.
+      attributes: node.attributes,
+      // Array of child node object ids.
+      children: [],
+      // The display name for the UI. This is either the lower casee of the node's tag
+      // name or the doctype string for a document type node.
+      displayName: node.nodeType === DOCUMENT_TYPE_NODE ? node.doctypeString : node.displayName,
+      // The computed display style property value of the node.
+      displayType: node.displayType,
+      // Whether or not the node has child nodes.
+      hasChildren: node.hasChildren,
+      // Whether or not the node has event listeners.
+      hasEventListeners: node.hasEventListeners,
+      // An unique NodeFront object id.
+      id,
+      // Whether or not the node is displayed. If a node has the attribute
+      // `display: none`, it is not displayed (faded in the markup view).
+      isDisplayed: node.isDisplayed,
+      // Whether or not the node is expanded.
+      isExpanded,
+      // Whether or not the node is an inline text child. NYI
+      isInlineTextChild: !!node.inlineTextChild,
+      // The namespace URI of the node. NYI
+      namespaceURI: node.namespaceURI,
+      // The object id of the parent node.
+      parentNodeId: parentNode?.objectId(),
+      // The pseudo element type.
+      pseudoType: node.pseudoType,
+      // The name of the current node.
+      tagName: node.tagName,
+      // The node's `nodeType` which identifies what the node is.
+      type: node.nodeType,
+      // The node's `nodeValue` which identifies the value of the current node.
+      value: node.getNodeValue(),
+    };
+  }
+
+  /**
+   * Collapses the given node in the markup tree.
+   *
+   * @param  {String} nodeId
+   *         The actor ID of the node to collapse.
+   */
+  collapseNode(nodeId) {
+    this.store.dispatch(updateNodeExpanded(nodeId, false));
+  }
+
+  /**
+   * Expands the given node in the markup tree.
+   *
+   * @param  {String} nodeId
+   *         The actor ID of the node to expand.
+   */
+  async expandNode(nodeId) {
+    await this.getChildren(this.nodes.get(nodeId));
+    this.tree[nodeId].isExpanded = true;
+    this.store.dispatch(updateTree(this.tree));
+  }
```

Comment:
I think if it were me, I'd move getChildren and some of this logic to redux actions, but up to you.